pkg/collector/component.go

```diff
@@ -0,0 +1,277 @@
+		portStr := ":" + strconv.Itoa(builderOpts.CollectorPort)
+		listener, err := net.Listen("tcp", portStr)
+		if err != nil {
+			c.logger.Fatal("Unable to start listening on channel", zap.Error(err))
+		}
+		c.logger.Info("Starting jaeger-collector TChannel server", zap.Int("port", builderOpts.CollectorPort))
+		c.logger.Warn("TChannel has been deprecated and will be removed in a future release")
+		c.tchServer.Serve(listener)
+	}
+
+	c.grpcServer, err = startGRPCServer(builderOpts, grpcHandler, strategyStore, c.logger)
+	if err != nil {
+		c.logger.Fatal("Could not start gRPC collector", zap.Error(err))
+	}
+
+	{
+		r := mux.NewRouter()
+		apiHandler := app.NewAPIHandler(jaegerBatchesHandler)
+		apiHandler.RegisterRoutes(r)
+
+		cfgHandler := clientcfgHandler.NewHTTPHandler(clientcfgHandler.HTTPHandlerParams{
+			ConfigManager: &clientcfgHandler.ConfigManager{
+				SamplingStrategyStore: strategyStore,
+				// TODO provide baggage manager
+			},
+			MetricsFactory:         c.metricsFactory,
+			BasePath:               "/api",
+			LegacySamplingEndpoint: false,
+		})
+		cfgHandler.RegisterRoutes(r)
+
+		recoveryHandler := recoveryhandler.NewRecoveryHandler(c.logger, true)
+		httpHandler := recoveryHandler(r)
+
+		httpPortStr := ":" + strconv.Itoa(builderOpts.CollectorHTTPPort)
+		c.logger.Info("Starting jaeger-collector HTTP server", zap.String("http-host-port", httpPortStr))
+
+		c.hServer = &http.Server{Addr: httpPortStr, Handler: httpHandler}
+		go func() {
+			if err := c.hServer.ListenAndServe(); err != nil {
+				if err != http.ErrServerClosed {
+					c.logger.Fatal("Could not launch service", zap.Error(err))
+				}
+			}
+			c.hCheck.Set(healthcheck.Unavailable)
+		}()
+
+		c.zkServer = zipkinServer(c.logger, builderOpts.CollectorZipkinHTTPPort, builderOpts.CollectorZipkinAllowedOrigins, builderOpts.CollectorZipkinAllowedHeaders, zipkinSpansHandler, recoveryHandler)
+		if c.zkServer != nil {
+			go func() {
+				if err := c.zkServer.ListenAndServe(); err != nil {
+					c.logger.Fatal("Could not launch Zipkin server", zap.Error(err))
+				}
+				c.hCheck.Set(healthcheck.Unavailable)
+			}()
+		}
+
+	}
+
+	return nil
+}
+
+// Close the component and all its underlying dependencies
+func (c *Component) Close() error {
+	c.grpcServer.GracefulStop() // gRPC
+	c.tchServer.Close()         // TChannel
+
+	// HTTP server
+	if c.hServer != nil {
+		timeout, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		err := c.hServer.Shutdown(timeout)
+		if err != nil {
+			c.logger.Error("Failed to stop the main HTTP server", zap.Error(err))
+		}
+		defer cancel()
+	}
+
+	// Zipkin server
+	if c.zkServer != nil {
+		timeout, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		err := c.zkServer.Shutdown(timeout)
+		if err != nil {
+			c.logger.Error("Failed to stop the Zipkin server", zap.Error(err))
+		}
+		defer cancel()
+	}
+
+	// by now, we shouldn't have any in-flight requests anymore
```

Comment:
Could there still be items in the collector queue? Or is it encapsulated behind the "writer" notion?

Reply:
Great question, I'll make sure to address this.